oak_channel/src/frame.rs

```diff
@@ -117,9 +115,11 @@ impl Framed {
             let body_length: usize = length
                 .checked_sub(BODY_OFFSET)
                 .expect("body length underflow");
-            let body = allocate_fn(body_length);
-            self.inner.read(body)?;
-            body
+            let mut body_buffer = Vec::with_capacity(body_length);
```

Comment:
I was trying really hard to get rid of any data copies in this code, and from my cursory look there can be three that happen here:
1. we load data into the Vec
2. we may need to copy _all_ of the data already in the `BytesMut` to make space for the new data
3. all the data in the Vec is copied into the `BytesMut`.

For (2), I think we should already know how much data we're expecting, so instead of `extend_from_slice` maybe use `put_slice` (w/ a call to `remaining_mut` to avoid a panic).
For (1) and (3), what we'd really want is a `body_length` size mutable slice direct from the `BytesMut` so that we could read directly into there instead of first reading into a `Vec` and then copying the data. But I don't know from the top of my head how to do it (or if it is indeed possible to do with the `BytesMut` API at all).

Reply:
+1 to @andrisaar comment.

Reply:
Done.

It seems BytesMut (or Rust in general?) does not allow to work with uninitialised memory. put_* is the only way to append. So we need two passes: (1) put_bytes to extend with zeros. (2) read into the slice zeroed out in (1). FWIW, coming from C++ I was looking for a way to avoid (1).

Reply:
Rust in general indeed doesn't like you to have access to uninitialised memory because of risk of undefined behaviour.

Our case is a bit special, as we're dealing with `u8`-s and we know that every (memory) value an `u8` could have would be a valid `u8` as we're dealing with raw bytes, after all.

In general, Rust has `MaybeUninit` to help with memory that may be uninitialised: https://doc.rust-lang.org/stable/core/mem/union.MaybeUninit.html

In our case, the "perfect" API would be that we allocate a buffer (and don't write anything into it), borrow a `&mut MaybeUninit<u8>` from there and then there would be some kind of `read_into` function that'd take the `&mut [MaybeUninit<u8>]` slice as an argument. (And maybe return a `&mut [u8]`, as after that the memory is _definitely_ intialised.)

However, I suspect such an API doesn't exist, and creating one would be annoying.